net: lib: nrf_cloud: REST lib error handling minor improvements (#6967)
nrfconnect/sdk-nrf: doc/nrf/libraries/networking/nrf_cloud_rest.rst

The nRF Cloud REST library enables applications to use `nRF Cloud's REST-based device API <nRF Cloud REST API_>`_.
This library is an enhancement to the :ref:`lib_nrf_cloud` library.

Usage
#####

> This library exposes a number of functions that allow access to nRF Cloud REST API endpoints.

Reviewer: allow --> provide

When called, these functions initiate an HTTP request to the nRF Cloud REST backend.
Various endpoints (documented `here <nRF Cloud REST API_>`_) are provided for such tasks as updating a device's shadow, sending and receiving device messages, interacting with location services, and so on.

> All of these functions operate within the context of a :c:struct:`nrf_cloud_rest_context` structure, which you must initialize and provide with your desired parameters.

Author: I have deviated slightly from your suggested changes on this one, are the deviations acceptable?

When each function returns, the context contains the status of the API call.

Authentication
**************

> Each HTTP/REST request must include a valid JSON Web Token, which serves as a proof of identity.

Reviewer 2: ```suggestion
Each HTTP/REST request must include a valid JSON Web Token (JWT) that serves as a proof of identity.
```

> These tokens are generated, usually, based on the `device credentials <nRF Cloud Security_>`_, and must be passed to the ``auth`` parameter of the :c:struct:`nrf_cloud_rest_context` structure.

Reviewer 2: ```suggestion
Usually, these tokens are generated based on the `device credentials <nRF Cloud Security_>`_, and must be passed to the ``auth`` parameter of the :c:struct:`nrf_cloud_rest_context` structure.
```

> JWT Tokens can be generated by calling :c:func:`nrf_cloud_jwt_generate` or using :ref:`lib_modem_jwt`.

Reviewer 2: ```suggestion
To generate a token, call the :c:func:`nrf_cloud_jwt_generate` function or use :ref:`lib_modem_jwt` library.
```

> Alternatively, if :kconfig:`NRF_CLOUD_REST_AUTOGEN_JWT` is enabled (along with its dependencies), the nRF Cloud REST library will generate a JWT automatically if one is not provided.

Reviewer 2: ```suggestion
Alternatively, if you have enabled the :kconfig:`NRF_CLOUD_REST_AUTOGEN_JWT` option (along with its dependencies), the nRF Cloud REST library generates a JWT automatically if one is not provided.
```

> Socket Management and Reuse

Reviewer 2: ```suggestion
Socket management and reuse
```

***************************

The socket used for each HTTP request is determined by the ``connect_socket`` parameter of the :c:struct:`nrf_cloud_rest_context` structure.

> If a valid socket file descriptor is passed, this socket is used. If -1 is passed, a socket will be created automatically.

Reviewer 2: ```suggestion
If a valid socket file descriptor is passed, this socket is used.
If ``-1`` is passed, a socket is created automatically.
```
Remember, each sentence on a separate line.

The selected socket can be reused depending on the passed-in value of the ``keep_alive`` property of the :c:struct:`nrf_cloud_rest_context` structure:

>  * If ``keep_alive`` is set to false, then the selected socket will be closed after each request, and the ``connect_socket`` property will be reset to -1.
>  * If ``keep_alive`` is set to true, the socket will remain open and ``connect_socket`` will remain unaltered, meaning the socket will be reused on the next API call.

Reviewer 2: ```suggestion
 * If you set ``keep_alive`` to false, the selected socket is closed after each request, and the ``connect_socket`` property is reset to ``-1``.
 * If  you set ``keep_alive`` to true, the socket remains open and ``connect_socket`` remains unaltered, meaning the socket is reused on the next API call.
```

> However, if ``keep_alive`` is set to true, care must be taken to ensure that the socket has not been closed externally (for example, due to inactivity) before sending each request.

Reviewer 2: ```suggestion
However, if  you set ``keep_alive`` to true, make sure that the socket has not been closed externally (for example, due to inactivity) before sending each request.
```

Otherwise, the request will be dropped.

There are two ways a timed out socket can manifest:

> * The socket may be closed immediately on timeout, causing `-ENOTCONN` to be returned immediately by the next REST request function call, without reaching nRF Cloud backend.

Reviewer: Overuse of "immediately"?

> * The socket may be left open, allowing the next REST request to reach nRF Cloud Backend, but the request will be ignored and the socket will be closed gracefully (FIN,ACK) with null HTTP response.

Reviewer: How do you know it reaches the backend?  It could have been closed by the carrier (without notice) due to a NAT timeout, in which case, the backend would never see it.  If the backend did receive it, it should process it and not return a null HTTP response.

Author: > How do you know it reaches the backend?
This is indicated by modem traces showing TLS alert response (with unknown content) originating from the endpoint IP. AFAIK only the backend could generate these encrypted alerts. Technically, the FIN,ACK also requires a valid sequence number which would be difficult for the carrier network to fake. The null HTTP response is a genuine response from back-end, and not generated at the carrier level.

> it should process it and not return a null HTTP response.

Officially, yes, all valid HTTP requests that reach an HTTP enpoint should receive valid HTTP responses. However AWS seems to violate this behavior if the socket being used has gone inactive, and instead gracefully closes the socket.

This does kind of depend on what is meant by backend. You could consider AWS as either the container for nRF Cloud Backend, or a necessary container that backend. It might be better to say "backend servers".

Reviewer: We don't (or shouldn't) refer to the backend per se in NCS documentation.  I think all of this is handled by AWS -- I don't think we host or manage the HTTP server or its responses.  I wonder if this error condition is specifically because the JWT expired?

Author: > We don't (or shouldn't) refer to the backend per se in NCS documentation.
> We don't (or shouldn't) refer to the backend per se in NCS documentation. 
Ah, well, I definitely think we need to somehow explain that there are two totally different ways a timeout can manifest, and that it's essentially random which of the two occurs. I'm not sure there's a way to explain what is happening without referring to the backend somehow. Would referring to it indirectly be acceptable? 

> I wonder if this error condition is specifically because the JWT expired?
Definitely not caused by the JWTs -- they are all valid -- a new one is generated for each and every request, even if the socket is being reused.

Reviewer: I don't think that saying "nRF Cloud backend" has any useful meaning or distinction to the customer from simply "nRF Cloud."  Such a distinction does not help them fix whatever is wrong.

Author: You're right, that is a vague term -- what about nRF Cloud REST endpoint server? Or just REST endpoint server? Or do you still think just saying "nRF Cloud" would be best?

Reviewer: IMO to a customer they are one and the same.

> * The socket may be closed immediately on timeout, causing `-ENOTCONN` to be returned by the next REST request function call, without reaching nRF Cloud.
> * The socket may be left open, allowing the next REST request to reach nRF Cloud, but the request will be ignored and the socket will be closed gracefully (FIN,ACK) with null HTTP response.
>   In this case, the called request function will return `-ESHUTDOWN`.

Reviewer 2: ```suggestion
* The socket may be closed immediately on timeout, causing ``-ENOTCONN`` to be returned by the next REST request function call, without reaching nRF Cloud.
* The socket may be left open, allowing the next REST request to reach nRF Cloud, but the request is ignored and the socket is closed gracefully (FIN,ACK) with null HTTP response.
  In this case, the called request function returns ``-ESHUTDOWN``.
```

See :ref:`nrf_cloud_rest_failure` for more details.

Request Success
***************
If the API call is successful, the context also contains the response data and length.
Some functions also have an optional ``result`` parameter.
If this parameter is provided, the response data is parsed into ``result``.

A number of functions automatically handle response data themselves:

* :c:func:`nrf_cloud_rest_pgps_data_get` - Pass response data to the P-GPS library :ref:`lib_nrf_cloud_pgps`.
* :c:func:`nrf_cloud_rest_agps_data_get` - Pass response data to the A-GPS library :ref:`lib_nrf_cloud_agps`.
* :c:func:`nrf_cloud_rest_fota_job_get` - If a FOTA job exists, :ref:`lib_fota_download` can perform the firmware download and installation. Call the :c:func:`nrf_cloud_rest_fota_job_update` function to report the status of the job.

.. _nrf_cloud_rest_failure:

Request Failure
***************

If an API call is unsuccessful, the called request function may return a variety of outputs:

* If the error happened at the socket level, the exact socket errno is returned.

>   (For instance, `-ENOTCONN` if the socket was closed, or was never opened before the request was made)

Reviewer: I think the parens should be removed and a period placed at the end.

* If the remote endpoint closes the connection gracefully without giving a response (a null HTTP response), `-ESHUTDOWN` is returned.
* If the remote endpoint responds with an unexpected HTTP status code (indicating request rejection), `-EBADMSG` is returned.

>   (Possible causes include, but are not limited to: bad endpoint, invalid request data, invalid JWT, so on)

Reviewer: same as above re parens

> * If the response body is empty and the request expects response data, `-ENODATA` will be returned.

Reviewer 2: ```suggestion
* If the response body is empty and the request expects response data, ``-ENODATA`` is returned.
```

> * If a heap allocation fails, `-ENOMEM` will be returned.

Reviewer 2: ```suggestion
* If a heap allocation fails, ``-ENOMEM`` is returned.
```

> * Request formatting errors will return `-ETXTBSY`

Reviewer 2: ```suggestion
* Request formatting errors return ``-ETXTBSY``.
```

> Some functions may have additional return values. These will be documented on the function itself.

Reviewer 2: ```suggestion
Some functions may have additional return values.
These are documented on the function itself.
```
